staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go

```diff
@@ -1670,3 +1691,515 @@ func TestCacheWatcherDrainingRequestedButNotDrained(t *testing.T) {
+				verifyNoEvents(t, w)
+				// add a pod that is greater than the storage's RV when the watch was started
+				err = cacher.watchCache.Add(makePod(103))
+				require.NoError(t, err, "failed to add a pod: %v")
+				verifyEvents(t, w, []watch.Event{{Type: watch.Added, Object: makePod(103)}})
+				verifyNoEvents(t, w)
+			},
+		},
+		{
+			name: "legacy, RV=0",
+			run: func(t *testing.T, cacher *Cacher, setStorageListMetaRV func(rv string)) {
+				// add some initial data
+				err := cacher.watchCache.Add(makePod(101))
+				require.NoError(t, err, "failed to add a pod: %v")
+				err = cacher.watchCache.Add(makePod(102))
+				require.NoError(t, err, "failed to add a pod: %v")
+				// set storage's RV to some future value, should not be used by this scenario
+				setStorageListMetaRV("105")
+				w, err := cacher.Watch(context.Background(), "pods/ns", storage.ListOptions{ResourceVersion: "0", Predicate: storage.Everything})
+				require.NoError(t, err, "failed to create watch: %v")
+				defer w.Stop()
+				// make sure get initial events
+				verifyEvents(t, w, []watch.Event{{Type: watch.Added, Object: makePod(101)}, {Type: watch.Added, Object: makePod(102)}})
+				verifyNoEvents(t, w)
+			},
+		},
+		{
+			name: "legacy, RV=unset",
+			run: func(t *testing.T, cacher *Cacher, setStorageListMetaRV func(rv string)) {
+				// add some initial data
+				err := cacher.watchCache.Add(makePod(101))
+				require.NoError(t, err, "failed to add a pod: %v")
+				err = cacher.watchCache.Add(makePod(102))
+				require.NoError(t, err, "failed to add a pod: %v")
+				// set storage's RV to some future value, should not be used by this scenario
+				setStorageListMetaRV("105")
+				w, err := cacher.Watch(context.Background(), "pods/ns", storage.ListOptions{Predicate: storage.Everything})
+				require.NoError(t, err, "failed to create watch: %v")
+				defer w.Stop()
+				// make sure get initial events
+				verifyEvents(t, w, []watch.Event{{Type: watch.Added, Object: makePod(101)}, {Type: watch.Added, Object: makePod(102)}})
+				verifyNoEvents(t, w)
+			},
+		},
+	}
+
+	for _, scenario := range scenarios {
+		t.Run(scenario.name, func(t *testing.T) {
+			storageListMetaResourceVersion := "100"
+			backingStorage := &dummyStorage{getListFn: func(_ context.Context, _ string, _ storage.ListOptions, listObj runtime.Object) error {
+				podList := listObj.(*example.PodList)
+				podList.ListMeta = metav1.ListMeta{ResourceVersion: storageListMetaResourceVersion}
+				return nil
+			}}
+
+			cacher, _, err := newTestCacher(backingStorage)
+			if err != nil {
+				t.Fatalf("falied to create cacher: %v", err)
+			}
+			defer cacher.Stop()
+			if err := cacher.ready.wait(); err != nil {
+				t.Fatalf("unexpected error waiting for the cache to be ready")
+			}
+			scenario.run(t, cacher, func(storageListMetaRV string) { storageListMetaResourceVersion = storageListMetaRV })
+		})
+	}
+}
+
+// sendInitialEvents=true, RV=""
+func TestCacheWatchSemanticsScenario1(t *testing.T) {
```

Comment:
Is this test really needed - I think this is now tested in the test above.

Reply:
yes, cruft, sorry, will remove.